Move score script context from SearchScript to its own class

martijnvg opened the pull request.

rjernst commented on server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java:
```diff
@@ -58,27 +58,26 @@ public DocIdSetIterator iterator() {
-        final SearchScript leafScript = script.newInstance(ctx);
+        final ScoreScript leafScript = script.newInstance(ctx);
         final CannedScorer scorer = new CannedScorer();
-        leafScript.setScorer(scorer);
```
I think we want to keep the scorer. Having the script access score through `get_score()` (which maps to `_score`) allows score to not be computed when it is not used. needs_score only gets us so far (whether it is used at all). With get_score, we could (although I'm not sure it is actually implemented this way yet) call the method lazily to bind _score.

martijnvg commented:
Thanks for reviewing! I've updated the pr and brought back setScorer() and added get_score() which is backed by a supplier that lazily computes the score if needed.
